### PR TITLE
Add throughputs to stage and job metrics

### PR DESCRIPTION
**Pull request by ccanel**
Modified the classes in metrics.py to calculate utilization and
effective throughput. Added a new metrics class: CpuMetrics. We cannot
calculate the disk read and disk write throughputs separately because
the disk counters that track read/write access time are not updated
live; instead, they are only updated when an access completes, which
leads to inaccuracies.

### CODE REVIEW

File: metrics.py [hunk 1]
```diff
@@ -55,21 +93,31 @@ def __init__(self, elapsed_millis, start_counters, end_counters):
...
     self.millis_writing = (end_counters.get(MILLIS_WRITING_KEY, 0.) -
       start_counters.get(MILLIS_WRITING_KEY, 0.))
+    self.total_io_millis = (end_counters.get(TOTAL_IO_MILLIS_KEY, 0.) -
+      start_counters.get(TOTAL_IO_MILLIS_KEY, 0.))
+    self.__calculate_util_and_throughput()
 
   def add_metrics(self, other_metrics):
     self.elapsed_millis += other_metrics.elapsed_millis
     self.bytes_read += other_metrics.bytes_read
     self.millis_reading += other_metrics.millis_reading
```
Comment (kayousterhout) [inline metrics.py:107]: Can you add a note by this and millis_writing to point out that they may not be accurate?

**Comment by kayousterhout** [comment]
Looks good other than the small comment! Did you sanity check these results with some logs you have and looking at the associated continuous monitor?

File: metrics.py (hunk 1 above)
Comment (ccanel) [inline metrics.py:107]: Done

**Comment by ccanel** [comment]
I compared the output to a continuous monitor, and the CPU, network, and disk metrics look reasonable. There is definitely something wrong with the GC time though. It looks like a lot of the minor GCs aren't getting counted. Which is weird because they show up in the continuous monitor and both codepaths capture the GC time in the same way. Will investigate more.

**Comment by kayousterhout** [comment]
Merged!

**Pull request closed without merging**